runtime/win32.c

```diff
@@ -1113,54 +1104,12 @@ void caml_init_os_params(void)
   QueryPerformanceFrequency(&frequency);
   /* Convert the frequency to the duration of 1 tick in ns */
   frequency.QuadPart = 1000000000LL / frequency.QuadPart;
-
-  /* Get the current time as accurately as we can.
-     GetSystemTimePreciseAsFileTime is available on Windows 8 / Server 2012+ and
-     gives <1us precision. For Windows 7 and earlier, which is only accurate to
-     10-100ms. */
-  pGetSystemTime =
-    (LPFN_GETSYSTEMTIME)GetProcAddress(GetModuleHandle(L"kernel32"),
-                                       "GetSystemTimePreciseAsFileTime");
-  if (!pGetSystemTime)
-    pGetSystemTime = GetSystemTimeAsFileTime;
-
-  /* Get the time and the performance counter. Get the performance counter first
-     to ensure no quantum effects */
-  QueryPerformanceCounter(&counter);
-  pGetSystemTime(&stamp);
-
-  now.LowPart = stamp.dwLowDateTime;
-  now.HighPart = stamp.dwHighDateTime;
-
-  /* Convert a FILETIME in 100ns ticks since 1 January 1601 to
-     ns since 1 Jan 1970. */
-  clock_offset.QuadPart =
-    ((now.QuadPart - 0x19DB1DED53E8000ULL) * 100);
-
-  /* Get the offset between QueryPerformanceCounter and
-     GetSystemTimePreciseAsFileTime in order to return a true timestamp, rather
-     than just a monotonic time source */
-  clock_offset.QuadPart -= (counter.QuadPart * frequency.QuadPart);
-
-  GetSystemTimePreciseAsFileTime(&stamp);
-  now.LowPart = stamp.dwLowDateTime;
-  now.HighPart = stamp.dwHighDateTime;
-  now.QuadPart *= 100;
 }
 
 int64_t caml_time_counter(void)
 {
-  static double clock_freq = 0;
-  static LARGE_INTEGER now;
+  LARGE_INTEGER now;
 
-  if (clock_freq == 0) {
-    LARGE_INTEGER f;
-    if (!QueryPerformanceFrequency(&f))
-      return 0;
-    clock_freq = (1000000000.0 / f.QuadPart);
-  };
-
-  if (!QueryPerformanceCounter(&now))
-    return 0;
-  return (int64_t)(now.QuadPart * clock_freq);
```

Comment:
Apologies for beating a dead horse, but the old code correctly uses floating-point to do the conversion, while the new code is just wrong in the way it uses integer arithmetic.  According to Web wisdom, typical values for the result of `QueryPerformanceFrequency` are in the 2e6...10e6  range, so by computing 1e9 / f with integer arithmetic, the ratio is off by up to 1% and your timings are wrong by as much.

Now if you keep the old code for `caml_time_counter`, the global `frequency` variable and its initialization in `caml_init_os_params` become useless.

One last nit: if `f` is a frequency, `1e9 / f` is a period, so `clock_freq` should probably be renamed `clock_period`.